src/library_atomic.js

```diff
@@ -0,0 +1,151 @@
+//   https://bugzilla.mozilla.org/show_bug.cgi?id=1467846
+// And at the time of writing, no other browser has it either.
+#if MIN_EDGE_VERSION < 91 || MIN_CHROME_VERSION < 91 || MIN_SAFARI_VERSION != TARGET_NOT_SUPPORTED || MIN_FIREFOX_VERSION != TARGET_NOT_SUPPORTED || ENVIRONMENT_MAY_BE_NODE
+  // Partially polyfill Atomics.waitAsync() if not available in the browser.
+  // Also polyfill for old Chrome-based browsers, where Atomics.waitAsync is
+  // broken until Chrome 91, see:
+  //   https://bugs.chromium.org/p/chromium/issues/detail?id=1167541
+  //   https://github.com/tc39/proposal-atomics-wait-async/blob/master/PROPOSAL.md
+  // This polyfill performs polling with setTimeout() to observe a change in the
+  // target memory location.
+  $polyfillWaitAsync__postset: `if (!Atomics.waitAsync || (typeof navigator !== 'undefined' && navigator.userAgent && jstoi_q((navigator.userAgent.match(/Chrom(e|ium)\\/([0-9]+)\\./)||[])[2]) < 91)) {
+  let __Atomics_waitAsyncAddresses = [/*[i32a, index, value, maxWaitMilliseconds, promiseResolve]*/];
+  function __Atomics_pollWaitAsyncAddresses() {
+    let now = performance.now();
+    let l = __Atomics_waitAsyncAddresses.length;
+    for (let i = 0; i < l; ++i) {
+      let a = __Atomics_waitAsyncAddresses[i];
+      let expired = (now > a[3]);
+      let awoken = (Atomics.load(a[0], a[1]) != a[2]);
+      if (expired || awoken) {
+        __Atomics_waitAsyncAddresses[i--] = __Atomics_waitAsyncAddresses[--l];
+        __Atomics_waitAsyncAddresses.length = l;
+        a[4](awoken ? 'ok': 'timed-out');
+      }
+    }
+    if (l) {
+      // If we still have addresses to wait, loop the timeout handler to continue polling.
+      setTimeout(__Atomics_pollWaitAsyncAddresses, 10);
+    }
+  }
+  #if ASSERTIONS && WASM_WORKERS
+    if (!ENVIRONMENT_IS_WASM_WORKER) err('Current environment does not support Atomics.waitAsync(): polyfilling it, but this is going to be suboptimal.');
+  #endif
+  /**
+   * @param {number=} maxWaitMilliseconds
+   */
+  Atomics.waitAsync = (i32a, index, value, maxWaitMilliseconds) => {
+    let val = Atomics.load(i32a, index);
+    if (val != value) return { async: false, value: 'not-equal' };
+    if (maxWaitMilliseconds <= 0) return { async: false, value: 'timed-out' };
+    maxWaitMilliseconds = performance.now() + (maxWaitMilliseconds || Infinity);
+    let promiseResolve;
+    let promise = new Promise((resolve) => { promiseResolve = resolve; });
+    if (!__Atomics_waitAsyncAddresses[0]) setTimeout(__Atomics_pollWaitAsyncAddresses, 10);
+    __Atomics_waitAsyncAddresses.push([i32a, index, value, maxWaitMilliseconds, promiseResolve]);
+    return { async: true, value: promise };
+  };
+}`,
+#endif
+
+  $polyfillWaitAsync__internal: true,
+  $polyfillWaitAsync__deps: ['$jstoi_q'],
+  $polyfillWaitAsync: () => {
+    // nop, used for its postset to ensure waitAsync() polyfill is included
+    // exaclyt once.
+    // Any function using Atomics.waitAsync should depend on this.
```

Comment:
Hmm, I suppose you are looking to fix a situation where only JS code would ever be doing Atomics.waitAsync()s, and Wasm code is doing all the wakes, hence the polyfill would never be emitted?

This situation looks like something that would be good to go in the Atomics related documentation?

Reply:
Reviewing this PR for these functional changes is a bit tricky, given the rename. Are there other functional differences?

Reply:
> Hmm, I suppose you are looking to fix a situation where only JS code would ever be doing Atomics.waitAsync()s, and Wasm code is doing all the wakes, hence the polyfill would never be emitted?
> 
> This situation looks like something that would be good to go in the Atomics related documentation?

I'm not quite sure I understand.   `Atomics.waitAsync` can only ever be performed by JS code, since its JS only function, right?  There is no wasm version of it.

This dependency just means the polyfill is only ever included when the library functions that call it are included.   Isn't that desirable?  Or would you rather the polyfill always be included?

Reply:
> Reviewing this PR for these functional changes is a bit tricky, given the code move. Are there other functional differences?

I'm pretty sure I didn't change anything functionally on the JS library side, just moved stuff around.

Reply:
> Hmm, I suppose you are looking to fix a situation where only JS code would ever be doing Atomics.waitAsync()s, and Wasm code is doing all the wakes, hence the polyfill would never be emitted?
> 
> This situation looks like something that would be good to go in the Atomics related documentation?

I can split out the `__deps` fix and land that first to make it more obvious what is going on here ..

Reply:
There is one change to the JS code here, in order to support running `emscripten_atomic_wait_async` under node:  I added runtimeKeepalivePush/runtimeKeepalivePop and callUserCallback to `emscripten_atomic_wait_async`.   This means that the runtime will not exit will async waits are running, but will exit cleanly once the last one is done.